packages/next-swc/crates/next-core/src/next_font/local/font_fallback.rs

```diff
@@ -52,3 +74,369 @@ pub(super) async fn get_font_fallbacks(
+    let mut font = Font::new(scope.read::<FontData>()?.table_provider(0)?)?.context(format!(
+        "Unable to read font metrics from font file at {}",
+        &main_descriptor.path,
+    ))?;
+
+    let az_avg_width = calc_average_width(&mut font);
+    let units_per_em = font
+        .head_table()?
+        .context(format!(
+            "Unable to read font scale from font file at {}",
+            &main_descriptor.path
+        ))?
+        .units_per_em as f64;
+
+    let fallback_avg_width = fallback_font.az_avg_width / fallback_font.units_per_em as f64;
+    let size_adjust = match az_avg_width {
+        Some(az_avg_width) => az_avg_width as f64 / units_per_em / fallback_avg_width,
+        None => 1.0,
+    };
+
+    Ok(FontAdjustment {
+        ascent: font.hhea_table.ascender as f64 / (units_per_em * size_adjust),
+        descent: font.hhea_table.descender as f64 / (units_per_em * size_adjust),
+        line_gap: font.hhea_table.line_gap as f64 / (units_per_em * size_adjust),
+        size_adjust,
+    })
+}
+
+fn calc_average_width(font: &mut Font<DynamicFontTableProvider>) -> Option<f32> {
+    let has_all_glyphs = AVG_CHARACTERS.chars().all(|c| {
+        font.lookup_glyph_index(c, allsorts::font::MatchingPresentation::NotRequired, None)
+            .0
+            > 0
+    });
+    if !has_all_glyphs {
+        return None;
+    }
+
+    Some(
+        font.map_glyphs(
+            AVG_CHARACTERS,
+            allsorts::tag::LATN,
+            allsorts::font::MatchingPresentation::NotRequired,
+        )
+        .iter()
+        .map(|g| font.horizontal_advance(g.glyph_index).unwrap())
+        .sum::<u16>() as f32
+            / AVG_CHARACTERS.len() as f32,
+    )
+}
+
+/// From https://github.com/vercel/next.js/blob/dbdf47cf617b8d7213ffe1ff28318ea8eb88c623/packages/font/src/local/pick-font-file-for-fallback-generation.ts#L59
+///
+/// If multiple font files are provided for a font family, we need to pick
+/// one to use for the automatic fallback generation. This function returns
+/// the font file that is most likely to be used for the bulk of the text on
+/// a page.
+///
+/// There are some assumptions here about the text on a page when picking the
+/// font file:
+/// - Most of the text will have normal weight, use the one closest to 400
+/// - Most of the text will have normal style, prefer normal over italic
+/// - If two font files have the same distance from normal weight, the thinner
+///   one will most likely be the bulk of the text
+fn pick_font_for_fallback_generation(
+    font_descriptors: &FontDescriptors,
+) -> Result<&FontDescriptor> {
+    match font_descriptors {
+        FontDescriptors::One(descriptor) => Ok(descriptor),
+        FontDescriptors::Many(descriptors) => {
+            let mut used_descriptor = descriptors
+                .first()
+                .context("At least one font is required")?;
+
+            for current_descriptor in descriptors.iter().skip(1) {
```

Comment:
Is this whole loop just a [min_by_key](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.min_by_key)?